### PR TITLE
Detect loop and isolation in DAG

### PR DESCRIPTION
**Pull request by Chiiizzzy**

### CODE REVIEW

File: towhee/dag/base_repr.py [hunk 1]
```diff
@@ -37,29 +36,28 @@ def name(self):
...
         if not isinstance(info, dict) or not essentials.issubset(info_keys):
-            logging.error(
-                'Info [%s] is not valid, lost attr [%s]', str(info), essentials - info_keys)
+            logging.error('Info [%s] is not valid, lack attr [%s]', str(info), essentials - info_keys)
             return False
         return True
 
     @staticmethod
-    def load_str(string: str) -> List[dict]:
+    def load_str(string: str) -> dict:
```
Comment (fzliu) [inline towhee/dag/base_repr.py:52]: This should be `typing.Dict`.

File: towhee/dag/graph_repr.py [hunk 2]
```diff
@@ -46,12 +45,125 @@ def operators(self) -> Dict[str, OperatorRepr]:
...
+                name = df['df']
+                if name not in dataframes:
+                    adj[name] = []
+                    flag[name] = 0
+                    dataframes.append(name)
+                adj[name].append(out)
+
+        # Recursive DFS
+        for df in dataframes:
+            status, msg = self.dfs(df, adj, flag, [])
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:109]: Since this is a static method, we should call `GraphRepr.dfs`.

File: towhee/dag/graph_repr.py [hunk 3]
```diff
@@ -46,12 +45,125 @@ def operators(self) -> Dict[str, OperatorRepr]:
...
+
+        for op in self._operators:
+            cur_in = {i['df'] for i in op.inputs}
+            cur_out = {i['df'] for i in op.outputs}
+
+            if not cur_in.intersection(out_df - cur_out) and not cur_out.intersection(in_df - cur_in):
+                iso_op.add(op.name)
+
+        status = bool(iso_op)
+        msg = '' if not iso_op else f'The DAG contains isolated operator(s) {iso_op}.'
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:157]: We should either log the error in this function or return the message to the caller, but not both.

File: towhee/dag/operator_repr.py [hunk 4]
```diff
@@ -30,10 +30,10 @@ class OperatorRepr(BaseRepr):
         iterator:
             This operator's iterator info.
     """
-
-    def __init__(self, name: str, function: str, init_args: Dict[str, any],
-                 inputs: List[Dict[str, any]], outputs: List[Dict[str, any]],
-                 iter_info: Dict[str, any]):
+    def __init__(
+        self, name: str, function: str, init_args: Dict[str, any], inputs: List[Dict[str, any]], outputs: List[Dict[str, any]], iter_info: Dict[str,
+                                                                                                                                                any]
```
Comment (fzliu) [inline towhee/dag/operator_repr.py:35]: There's too much whitespace here, you can insert a newline after any `,` in the parameter list.

File: towhee/dag/operator_repr.py [hunk 5]
```diff
@@ -134,11 +134,7 @@ def from_dict(info: Dict) -> 'OperatorRepr':
...
-                                        'inputs', 'outputs', 'iter_info'}):
-            raise ValueError(
-                'Invalid operator data.'
-            )
+        if not BaseRepr.is_valid(info, {'name', 'init_args', 'function', 'inputs', 'outputs', 'iter_info'}):
+            raise ValueError('Invalid operator info.')
 
-        return OperatorRepr(info['name'], info['function'], info['init_args'],
-                            info['inputs'], info['outputs'], info['iter_info'])
+        return OperatorRepr(info['name'], info['function'], info['init_args'], info['inputs'], info['outputs'], info['iter_info'])
```
Comment (fzliu) [inline towhee/dag/operator_repr.py:140]: Does this pass our line length requirements? It's fairly long.

File: towhee/dag/graph_repr.py [hunk 6]
```diff
@@ -46,12 +45,125 @@ def operators(self) -> Dict[str, OperatorRepr]:
...
+            status, msg = self.dfs(df, adj, flag, [])
+            if status:
+                logging.error(msg)
+                return status, msg
+
+        return False, ''
+
+    def has_isolated_df(self):
+        non_iso_df = set()
+        iso_df = set()
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:118]: It's more efficient to add all dataframes into the set first, then remove them as we detect that they are not isolated. The same goes for operators.

File: towhee/dag/operator_repr.py (hunk 5 above)
Comment (Chiiizzzy) [reply towhee/dag/operator_repr.py:140]: fixed

File: towhee/dag/operator_repr.py (hunk 4 above)
Comment (Chiiizzzy) [reply towhee/dag/operator_repr.py:35]: fixed

File: towhee/dag/graph_repr.py (hunk 3 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:157]: fixed

File: towhee/dag/graph_repr.py (hunk 2 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:109]: fixed

File: towhee/dag/base_repr.py (hunk 1 above)
Comment (Chiiizzzy) [reply towhee/dag/base_repr.py:52]: fixed

File: towhee/dag/graph_repr.py (hunk 6 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:118]: fixed

File: towhee/dag/graph_repr.py [hunk 7]
```diff
@@ -46,12 +44,118 @@ def operators(self) -> Dict[str, OperatorRepr]:
...
+            if status:
+                return status, msg
+
+        return False, ''
+
+    def has_isolated_df(self):
+        iso_df = set()
+
+        for df in self._dataframes:
+            iso_df.add(df.name)
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:118]: You can change this into a set comprehension (more Pythonic):
`iso_df = {df.name for df in self._dataframes}`

File: towhee/dag/graph_repr.py [hunk 8]
```diff
@@ -46,12 +44,115 @@ def operators(self) -> Dict[str, OperatorRepr]:
...
+            return False, ''
+
+        in_df = set()
+        out_df = set()
+        iso_op = set()
+
+        for op in self._operators:
+            for i in op.inputs:
+                in_df.add(i['df'])
+            out_df.add(op.outputs[0]['df'])
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:138]: We can also use set comprehensions here.

**Pull request merged**